packages/cache/src/lib/CacheMiddleware.ts

```diff
@@ -0,0 +1,672 @@
+  public async [Method.Filter](payload: Payloads.Filter.ByValue<StoredValue>): Promise<Payloads.Filter.ByValue<StoredValue>>;
+
+  @PreProvider()
+  public async [Method.Filter](payload: Payloads.Filter<StoredValue>): Promise<Payloads.Filter<StoredValue>> {
+    const { provider: cache } = this.context;
+    const { hook, path, type, value } = payload;
+
+    payload.data = {};
+
+    if (hook && isFilterByHookPayload(payload)) {
+      const cacheHook = async (value: CacheMiddleware.Document<StoredValue>, key: string) => {
+        if (await this.checkNotExpired(value, key)) {
+          if (await hook(value.value, key)) {
+            payload.data ??= {};
+            payload.data[key] = value.value;
+          }
+        }
+
+        const bypassPayload = await this.provider[Method.Get]({ method: Method.Get, key, path: [], errors: [] });
+
+        if (isPayloadWithData<StoredValue>(bypassPayload)) {
+          const { data } = bypassPayload;
+
+          if (await hook(data, key)) {
+            payload.data ??= {};
+            payload.data[key] = data;
+          }
+        }
+
+        return false;
+      };
+
+      const { errors } = await cache[Method.Filter]({ method: Method.Filter, hook: cacheHook, type, path: [], errors: [] });
+
+      payload.errors = [...payload.errors, ...errors];
+    } else {
+      const filterPayload = await cache[Method.Filter]({
+        method: Method.Filter,
+        type,
+        data: {},
+        value,
+        path: ['value', ...(path ?? [])],
+        errors: []
+      });
+
+      if (isPayloadWithData<CacheMiddleware.Document<StoredValue>>(filterPayload)) {
+        const { data } = filterPayload;
+        const final: { [key: string]: StoredValue } = {};
+
+        for (const [key, val] of Object.entries(data)) {
+          if (await this.checkNotExpired(val, key)) {
+            final[key] = val.value;
+          } else {
+            const bypassPayload = await this.provider[Method.Get]({ method: Method.Get, key, path: path ?? [], errors: [] });
+
+            if (isPayloadWithData<StoredValue>(bypassPayload)) {
+              const { data } = bypassPayload;
+
+              final[key] = data;
+            }
+          }
+        }
+
+        payload.data = final;
+      }
+
+      payload.errors = [...payload.errors, ...filterPayload.errors];
+    }
+
+    return payload;
+  }
+
+  public async [Method.Find](payload: Payloads.Find.ByHook<StoredValue>): Promise<Payloads.Find.ByHook<StoredValue>>;
+  public async [Method.Find](payload: Payloads.Find.ByValue<StoredValue>): Promise<Payloads.Find.ByValue<StoredValue>>;
+  @PreProvider()
+  public async [Method.Find](payload: Payloads.Find<StoredValue>): Promise<Payloads.Find<StoredValue>> {
+    const { provider: cache } = this.context;
+    const { hook, path, type, value } = payload;
+
+    if (hook && isFindByHookPayload(payload)) {
```

Comment:
```suggestion
    if (isFindByHookPayload(payload)) {
```
Destructure `hook` from `payload` inside this scope.